jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java

```diff
@@ -0,0 +1,184 @@
+import org.xwiki.model.reference.LocalDocumentReference;
+import org.xwiki.model.reference.WikiReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+import org.xwiki.query.QueryManager;
+import org.xwiki.user.UserReference;
+import org.xwiki.user.internal.document.DocumentUserReference;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * Listener to handle migration of authentication in JIRA configuration.
+ *
+ * @version $Id$
+ * @since 11.0.0
+ */
+@Singleton
+@Component
+@Named(JIRAConfigurationMigrator.HINT)
+public class JIRAConfigurationMigrator extends AbstractEventListener
+{
+    /**
+     * Main name of the listener.
+     */
+    public static final String HINT = "JIRAConfigMigrationListener";
+
+    private static final String JIRA = "JIRA";
+
+    private static final String PROPERTY_USERNAME = "username";
+
+    private static final String PROPERTY_PASSWORD = "password";
+
+    private static final String PROPERTY_ID = "id";
+
+    private static final String PROPERTY_AUTHENTICATION_TYPE = "authenticationType";
+
+    private static final String DOC_SAVE_MESSAGE = "Migrate configration";
+
+    private static final LocalDocumentReference MIGRATION_STATUS_DOCUMENT_REFERENCE =
+        new LocalDocumentReference(JIRA, "JIRAConfigAuthMigration");
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    @Inject
+    private Provider<QueryManager> queryManagerProvider;
+
+    @Inject
+    @Named("currentmixed")
+    private DocumentReferenceResolver<String> currentMixedReferenceResolver;
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Public constructor.
+     */
+    public JIRAConfigurationMigrator()
+    {
+        super(HINT, List.of(new ApplicationReadyEvent(), new JobFinishedEvent("install")));
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        migrate();
+    }
+
+    /**
+     * Run the data migration for JIRA server config.
+     */
+    public void migrate()
+    {
+        XWikiContext context = contextProvider.get();
+        XWiki xwiki = context.getWiki();
+        DocumentReference migrationDocumentRef =
+            new DocumentReference(MIGRATION_STATUS_DOCUMENT_REFERENCE, new WikiReference(context.getWikiId()));
+
+        try {
+            if (!xwiki.exists(migrationDocumentRef, context)) {
+                logger.info("Running migration of JIRA configuration");
+
+                XWikiDocument basicAuthDoc =
+                    xwiki.getDocument(BasicAuthJIRAAuthenticatorFactory.BASIC_AUTH_CONFIG_REFERENCE, context).clone();
+                XWikiDocument jiraServerDoc =
+                    xwiki.getDocument(new LocalDocumentReference(JIRA, "JIRAConfig"), context).clone();
```

Comment:
why do you keep cloning your docs?

Reply:
Because I suppose we should always clone it if we make modification on.